From 9caa19291e3fabc3e00319f3901393aba9d8372a Mon Sep 17 00:00:00 2001
From: seirra <sophietheopossum@yandex.com>
Date: Sun, 18 Oct 2020 21:22:13 +0100
Subject: [PATCH] remove architecture check, do not use --login

---
 bin/smartgit.sh | 25 +++++++------------------
 1 file changed, 7 insertions(+), 18 deletions(-)

diff --git a/bin/smartgit.sh b/bin/smartgit.sh
index c09e052..c5e32a8 100755
--- a/bin/smartgit.sh
+++ b/bin/smartgit.sh
@@ -1,12 +1,9 @@
-#!/bin/bash --login
+#!/bin/bash
 #
 # Editing this script should not be required.
 #
 # To specify additional VM options, add them to smartgit.vmoptions files.
 
-APP_NAME=SmartGit
-APP_LOWER=smartgit
-
 parseVmOptions() {
   if [ -f $1 ]; then
     while read LINE || [[ -n "$LINE" ]]; do
@@ -27,7 +24,7 @@ parseVmOptions() {
 
 echoJreConfigurationAndExit() {
     mkdir --parents $APP_CONFIG_DIR
-    touch $APP_CONFIG_DIR/$APP_LOWER.vmoptions
+    touch $APP_CONFIG_DIR/smartgit.vmoptions
     exit 1
 }
 
@@ -39,14 +36,6 @@ case "$BASH" in
         ;;
 esac
 
-
-# check system architecture
-ARCHITECTURE=`uname -m`
-if [ "$ARCHITECTURE" != "x86_64" ] && [ "$ARCHITECTURE" != "aarch64" ] ; then
-    echo "$APP_NAME is not supported any more on 32-bit systems."
-    exit 1
-fi
-
 # Resolve the location of the SmartGit installation.
 # This includes resolving any symlinks.
 PRG=$0
@@ -69,13 +58,13 @@ APP_BIN=`pwd`
 cd "${oldpwd}";
 unset oldpwd
 
-APP_CONFIG_DIR=${XDG_CONFIG_HOME:-$HOME/.config}/$APP_LOWER
+APP_CONFIG_DIR=${XDG_CONFIG_HOME:-$HOME/.config}/smartgit
 
 APP_HOME=`dirname "$APP_BIN"`
 APP_JAVA_HOME=$SMARTGIT_JAVA_HOME
-parseVmOptions $APP_BIN/$APP_LOWER.vmoptions
-parseVmOptions $HOME/.$APP_LOWER/$APP_LOWER.vmoptions
-parseVmOptions $APP_CONFIG_DIR/$APP_LOWER.vmoptions
+parseVmOptions $APP_BIN/smartgit.vmoptions
+parseVmOptions $HOME/.smartgit/smartgit.vmoptions
+parseVmOptions $APP_CONFIG_DIR/smartgit.vmoptions
 
 # Determine Java Runtime
 _JAVA_EXEC="java"
@@ -90,7 +79,7 @@ if [ "$APP_JAVA_HOME" != "" ] ; then
     else
         echo "Warning: $_TMP does not exist"
     fi
-elif [ -e "$APP_HOME/jre/bin/java" ] && [ "$ARCHITECTURE" == "x86_64" ] ; then
+elif [ -e "$APP_HOME/jre/bin/java" ] && [ `uname -m` == "x86_64" ] ; then
     _JAVA_EXEC="$APP_HOME/jre/bin/java"
 fi
 
-- 
2.26.2

