From de4fe3f45f2f5d0652a149395429d0fdb1a982ef Mon Sep 17 00:00:00 2001
From: seirra <sophietheopossum@yandex.com>
Date: Sun, 18 Oct 2020 21:23:27 +0100
Subject: [PATCH] bump max memory

---
 bin/smartgit.sh | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/bin/smartgit.sh b/bin/smartgit.sh
index 13e7070..8430405 100755
--- a/bin/smartgit.sh
+++ b/bin/smartgit.sh
@@ -93,4 +93,4 @@ fi
 
 APP_PATH="$PATH$SMARTGIT_PATH"
 
-(export PATH="$APP_PATH"; java -XX:+UseG1GC -XX:MaxGCPauseMillis=100 -XX:InitiatingHeapOccupancyPercent=25 -Xmx1024m -Xss2m -Xverify:none -XX:MaxJavaStackTraceDepth=1000000 -Dsun.io.useCanonCaches=false -XX:ErrorFile=$APP_CONFIG_DIR/20.2/hs_err_pid%p.log $_VM_PROPERTIES -jar "$APP_HOME/lib/bootloader.jar" "$@")
+(export PATH="$APP_PATH"; java -XX:+UseG1GC -XX:MaxGCPauseMillis=100 -XX:InitiatingHeapOccupancyPercent=25 -Xmx3072m -Xss2m -Xverify:none -XX:MaxJavaStackTraceDepth=1000000 -Dsun.io.useCanonCaches=false -XX:ErrorFile=$APP_CONFIG_DIR/20.2/hs_err_pid%p.log $_VM_PROPERTIES -jar "$APP_HOME/lib/bootloader.jar" "$@")
-- 
2.29.2

